lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -2262,4 +2341,539 @@ public void onError(ErrorInfo reason) {
+			channel1.presence.enterClient(testClientId1);
+			new PresenceWaiter(Action.enter, channel1).waitFor(1);
+
+			Channel channel2 = ably2.channels.get(channelName);
+			assertFalse("Verify SYNC is not complete yet", channel2.presence.syncComplete());
+			channel2.attach();
+			/* Wait for the SYNC to complete */
+			new PresenceWaiter(Action.present, channel2).waitFor(1);
+			try {
+				channel2.presence.get(true);
+			} catch (InterruptedException e) {}
+			/* Initial SYNC should be complete at this point */
+			assertTrue("Verify SYNC is complete", channel2.presence.syncComplete());
+		} catch (AblyException e) {
+			fail("Unexpected exception");
+		} finally {
+			if (ably1 != null)
+				ably1.close();
+			if (ably2 != null)
+				ably2.close();
+		}
+	}
+
+	/**
+	 * Enter client without permission to do so, check exception
+	 * Tests RTP8h
+	 */
+	@Test
+	public void presence_enter_without_permission() throws AblyException {
+		String channelName = "presence_enter_without_permission" + testParams.name;
+		AblyRealtime ably = null;
+
+		try {
+			/* init ably for token */
+			ClientOptions optsForToken = createOptions(testVars.keys[0].keyStr);
+			final AblyRest ablyForToken = new AblyRest(optsForToken);
+
+			/* get first token */
+			Auth.TokenParams tokenParams = new Auth.TokenParams();
+			Capability capability = new Capability();
+			capability.addResource(channelName, "publish");	/* no presence permission! */
+			tokenParams.capability = capability.toString();
+			tokenParams.clientId = testClientId1;
+
+			Auth.TokenDetails token = ablyForToken.auth.requestToken(tokenParams, null);
+			assertNotNull("Expected token value", token.token);
+
+			ClientOptions opts = createOptions();
+			opts.clientId = testClientId1;
+			opts.tokenDetails = token;
+			ably = new AblyRealtime(opts);
+
+			Channel channel = ably.channels.get(channelName);
+			channel.attach();
+
+			ChannelWaiter channelWaiter = new ChannelWaiter(channel);
+			channelWaiter.waitFor(ChannelState.attached);
+
+			final boolean[] flags = new boolean[] {
+					false, /* callback called */
+					false  /* success? */
+			};
+			channel.presence.enterClient(testClientId1, null, new CompletionListener() {
+				@Override
+				public void onSuccess() {
+					synchronized (flags) {
+						flags[0] = true;  flags[1] = true;
+						flags.notify();
+					}
+				}
+
+				@Override
+				public void onError(ErrorInfo reason) {
+					synchronized (flags) {
+						flags[0] = true;  flags[1] = false;
+						flags.notify();
+					}
+				}
+			});
+
+			try {
+				synchronized (flags) {
+					while (!flags[0])
+						flags.wait();
+				}
+			} catch (InterruptedException e) {}
+
+			assertTrue("Verify completion callback was called", flags[0]);
+			assertFalse("Verify enter client failed", flags[1]);
+		} finally {
+			if (ably != null)
+				ably.close();
+		}
+	}
+
+	/**
+	 * Enter wrong client (mismatching one set in the token), check exception
+	 * Tests RTP8i
```

Comment:
This is the wrong test, it stats `If the Ably service determines that the client is unidentified, a NACK is sent to the client resulting in an error`??  You are testing client ID mismatch instead of an unidentified client